compiler/hash-attrs/src/target.rs

```diff
@@ -0,0 +1,292 @@
+        const EnumDef = 1 << 15;
+
+        /// An enum variant.
+        const EnumVariant = 1 << 16;
+
+        /// A function definition, regardless of the position.
+        const FnDef = 1 << 17;
+
+        /// A type.
+        const Ty = 1 << 18;
+
+        /// A type argument, whilst this has similar implications to
+        /// a type, it adds additional context that this is a type
+        /// argument.
+        const TyArg = 1 << 19;
+
+        /// A parameter in a struct, enum, tuple or function definition.
+        const Param = 1 << 20;
+
+        /// A field in a struct or enum.
+        const Field = 1 << 21;
+
+        /// A match branch.
+        const MatchCase = 1 << 22;
+
+        /// A pattern.
+        const Pat  = 1 << 23;
+
+        /// A pattern argument.
+        const PatArg  = 1 << 24;
+
+        /// A general item definition e.g. `struct`, `enum`, `impl`, `mod` and `fn`.
+        const Item = Self::StructDef.bits() | Self::EnumDef.bits() | Self::FnDef.bits() | Self::TyFnDef.bits() | Self::ImplDef.bits() | Self::ModDef.bits();
+    }
+}
+
+impl AttrTarget {
+    /// Classify the given [ast::Expr] into a [AttrTarget].
+    pub fn classify_expr(expr: &ast::Expr) -> Self {
+        match expr {
+            ast::Expr::ConstructorCall(_) => AttrTarget::ConstructorCall,
+            ast::Expr::Macro(_) => AttrTarget::MacroInvocation,
+            ast::Expr::Unsafe(_) => AttrTarget::Unsafe,
+            ast::Expr::Lit(_) => AttrTarget::Lit,
+            ast::Expr::Block(ast::BlockExpr { data: block }) => match block.body() {
+                ast::Block::Loop(_) | ast::Block::While(_) | ast::Block::For(_) => AttrTarget::Loop,
+                ast::Block::Match(_) | ast::Block::If(_) => AttrTarget::Match,
+                ast::Block::Body(_) => AttrTarget::Block,
+            },
+            ast::Expr::Import(_) => AttrTarget::Import,
+            ast::Expr::StructDef(_) => AttrTarget::StructDef,
+            ast::Expr::EnumDef(_) => AttrTarget::EnumDef,
+            ast::Expr::TyFnDef(_) => AttrTarget::TyFnDef,
+            ast::Expr::TraitDef(_) => AttrTarget::Item,
+            ast::Expr::ImplDef(_) => AttrTarget::ImplDef,
+            ast::Expr::ModDef(_) => AttrTarget::ModDef,
+            ast::Expr::FnDef(_) => AttrTarget::FnDef,
+            ast::Expr::Ty(_) => AttrTarget::Ty,
+
+            // If this is a declaration, we have to recurse into the subject...
+            ast::Expr::Declaration(ast::Declaration { value: Some(value), .. }) => {
+                AttrTarget::classify_expr(value.body())
+            }
+            _ => AttrTarget::Expr,
+        }
+    }
+}
+
+impl fmt::Display for AttrTarget {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let item_count = self.bits().count_ones();
+
+        // We will collect all of the allowed argument kinds into a vector
+        let mut allowed_argument_kinds = Vec::with_capacity(item_count as usize);
+
+        for kind in (*self).iter() {
```

Comment:
Map?

Reply:
I don't think that is a thing

Reply:
I mean you can just do .iter().map() and not have to worry about dealing with vecs/pushing/reserving, but I guess it's fine either way

Reply:
Yeah I see now, but since this isn't quite like an `enum` so the match will always